### DISCUSSION TITLE
Obtaining parent/federated model from Bis.RepositoryLink class for each element

### DISCUSSION THREAD

**Discussion by nn15296** [iTwin/itwinjs-core#2893] [Q&A]
Hi

I'm querying elements from the Bis.GeometricElement3d class, and linking them to their projectwise models via the Bis.RepositoryLink class. I'd also like to obtain the parent federated model that each element is associated with as shown in the picture below:

![image](https://user-images.githubusercontent.com/33571096/145619720-010eafd2-1be5-4cf7-b690-017e670efae8.png)

I'm currently using the following ECSQL query:

```sql
SELECT
a.ecInstanceId
link.userLabel as model
FROM
Bis.GeometricElement3d a
JOIN
Bis.ExternalSourceAspect b ON a.ecInstanceId = b.element.id
JOIN
Bis.ExternalSource es ON b.source.id = es.ecInstanceId
JOIN
Bis.RepositoryLink link ON link.ecInstanceId = es.repository.id
```

Is there any way to obtain the federated model from the Bis.RepositoryLink class for each element that I'm querying from Bis.GeometricElement3d?

Thanks for the help.

**Comment by nn15296** [comment]
@ramanujam-raman any ideas?

**Comment by BaoTon** [comment]
Hi Nathan,

I think you'll need to join bis.elementscopesexternalsourceidentifier. Please let me know if this query works for you:

SELECT link.userlabel AS fileSource,
       elem3d.*
FROM biscore.geometricelement3d elem3d
LEFT JOIN bis.element elem ON elem3d.model.id = elem.ecinstanceid
INNER JOIN bis.externalsourceaspect AS asp ON asp.element.id = elem.ecinstanceid
INNER JOIN bis.elementscopesexternalsourceidentifier AS rel ON asp.ecinstanceid = rel.targetecinstanceid
INNER JOIN bis.repositorylink LINK ON link.ecinstanceid = rel.sourceecinstanceid

**Reply by nn15296** [reply]
Thanks for your reply! Unfortunately this still gives me the lowest level model, not the parent federated model.

**Reply by nn15296** [reply]
If you take the parent of the parent of that model, and then get the codevalue (as shown by f.codeValue in the first set of SQL below), then you get the federated parent model. But all the d.codeValue and e.codeValue just give you the lowest level model which is strange.

Because the first set of SQL below doesn't use the repository link, it doesn't provide me with the meta data from ProjectWise. When I try linking it via repository link as shown in the second set of SQL below, I don't get any results. I also don't get the right results if I try joining the bis.externalsourceaspect to d.ecInstanceId or e.ecInstanceId

```sql
SELECT
    d.codeValue AS model,
    e.codeValue AS parentOfModel,
    f.codeValue AS parentOfParentOfModel
FROM
    bis.geometricelement3d a
JOIN
    Bis.Element d ON a.model.id = d.ecInstanceId
JOIN
    Bis.Element e ON d.parent.id = e.ecInstanceId
JOIN
    Bis.Element f ON e.parent.id = f.ecInstanceId
```

```sql
SELECT
    link.userLabel AS federatedModel,
    link.jsonProperties AS PWMetaData
FROM
    bis.geometricelement3d a
JOIN
    Bis.Element d ON a.model.id = d.ecInstanceId
JOIN
    Bis.Element e ON d.parent.id = e.ecInstanceId
JOIN
    Bis.Element f ON e.parent.id = f.ecInstanceId
INNER JOIN 
    bis.externalsourceaspect AS asp ON asp.element.id = f.ecinstanceid
INNER JOIN
    bis.elementscopesexternalsourceidentifier AS rel ON asp.ecinstanceid = rel.targetecinstanceid
INNER JOIN
    bis.repositorylink LINK ON link.ecinstanceid = rel.sourceecinstanceid
```

**Reply by JHolland98** [reply]
I'm having the same issue @BaoTon

**Reply by BaoTon** [reply]
@swwilson-bsi Is there a way to link element parent id with bis.repositorylink to get the federated model?

**Reply by JHolland98** [reply]
@BaoTon thanks for the reply. @swwilson-bsi do you have an update on this? Thanks
